### PR TITLE
internal/v5: implement meta/resources endpoint

### PR DESCRIPTION
**Pull request by rogpeppe**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1724/
Test PASSed.

**Comment by mhilton** [comment]
:+1:

File: internal/v5/resources.go [hunk 1]
```diff
@@ -174,3 +135,58 @@ func validResourceName(name string) bool {
...
+		Revision:    -1,
+		Type:        meta.Type.String(),
+		Path:        meta.Path,
+		Description: meta.Description,
+	}
+	if doc.BlobHash == "" {
+		// No hash implies that there is no file (the entry
+		// is just a placeholder), so we don't fill in
+		// blob details.
+		logger.Infof("no blobhash for resource %q", doc.Name)
```
Comment (frankban) [inline internal/v5/resources.go]: .Debug?

File: internal/v5/resources.go [hunk 2]
```diff
@@ -174,3 +135,58 @@ func validResourceName(name string) bool {
...
+
+func fromResourceDoc(doc *mongodoc.Resource, resources map[string]resource.Meta) (*params.Resource, error) {
+	meta, ok := resources[doc.Name]
+	if !ok {
+		// Should never happen.
+		return nil, errgo.Newf("resource %q returned by ListResources but not found in metadata", doc.Name)
+	}
+	r := &params.Resource{
+		Name:        doc.Name,
+		Revision:    -1,
```
Comment (frankban) [inline internal/v5/resources.go:172]: Out of curiosity, isn't doc.Revision -1 in the case there is no file? Is this like this just to be more explicit?

File: internal/charmstore/resources.go [hunk 3]
```diff
@@ -58,6 +58,9 @@ func (sorted resourcesByName) Less(i, j int) bool {
 // revision for each resource. Otherwise it holds the revisions declared
 // when the charm/channel pair was published.
 func (s *Store) ListResources(entity *mongodoc.Entity, channel params.Channel) ([]*mongodoc.Resource, error) {
+	if channel == params.NoChannel {
+		return nil, errgo.Newf("no channel specified")
```
Comment (frankban) [inline internal/charmstore/resources.go:62]: Do we need a test for this case?

**Comment by frankban** [comment]
:+1: with a couple of questions. Thank you!

File: internal/v5/resources.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v5/resources.go:172]: Yes it is.

File: internal/v5/resources.go (hunk 1 above)
Comment (rogpeppe) [inline internal/v5/resources.go]: removed

File: internal/charmstore/resources.go (hunk 3 above)
Comment (rogpeppe) [inline internal/charmstore/resources.go:62]: added a test

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1725/
Test PASSed.

**Pull request merged**